common/stream-ratelimiter/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+//! `RateLimiter` converts any [`futures::stream::Stream`] into a rate-limited
+//! stream, allowing only a certain number of elements to be polled in a given time interval.
+//!
+//! RateLimiter uses the token-bucket algorithm for flow control. RateLimiter acquires a token/permit
+//! before polling any element from the underlying stream. If there are no more tokens, it must wait
+//! for new tokens to be added back through a periodic refill process.
+//!
+//! The token-bucket is in turn implemented using a futures-aware semapahore. The semaphore is
+//! initialized with some capacity, and the periodic refill process adds back permits removed
+//! during the previous interval.
+//!
+//! If fine-grained flow control is desired, the refill interval should be kept small.
+
+use futures::{
+    future::{Future, FutureExt},
+    ready,
+    stream::{Fuse, Stream, StreamExt},
+};
+use futures_semaphore::Semaphore;
+use pin_project::pin_project;
+use std::{mem::ManuallyDrop, pin::Pin, task, task::Poll, time::Duration};
+use tokio::time::{interval, Interval};
+
+/// Config parameters for a rate-limiter.
+/// `capacity`: Max elements allowed in an interval.
+/// `interval`: Granular duration within which control flow is desired.
+#[derive(Debug)]
+pub struct RateLimitConfig {
+    refill_interval: Duration,
+    capacity: usize,
+}
+
+impl RateLimitConfig {
+    pub fn new(refill_interval: Duration, capacity: usize) -> RateLimitConfig {
+        RateLimitConfig {
+            refill_interval,
+            capacity,
+        }
+    }
+}
+
+pub trait RateLimit: Stream {
+    fn ratelimit(self, config: RateLimitConfig) -> RateLimiter<Self>
+    where
+        Self: Sized,
+    {
+        RateLimiter::new(config, self)
+    }
+}
+
+impl<T: ?Sized> RateLimit for T where T: Stream {}
```

Comment:
Depending on the sort of UX you're trying to get you might be able to simplify the number of types that a user needs to use to construct this. A couple options:
1) Drop the RateLimit trait and just have a `new` function which looks like:
```rust
pub fn new<S: Stream>(stream: S, capacity: usize, refill_interval: Duration) -> RateLimit
```
2) Fold the `RateLimitConfig` object into the ratelimit trait so that the new fn looks like:
```rust
fn ratelimit(self, capacity: usize, refill_interval: Duration) -> RateLimiter<Self>
```

Reply:
I really do like the `RateLimit` trait since it allows to simply call `.ratelimit()` on any stream and construct a chain, making the overall use more ergonomic. Removing the RateLimitConfig struct seems reasonable so going forward with that. Thanks!